### PR TITLE
feat(nargo)!: save program ABI alongside ACIR

### LINKED ISSUE #897: Save Noir ABI to disk
# Problem

It is not convenient to arrange the initial witness for a precompiled ACIR. At present, the Noir ABI is recompiled from source and held in memory for all nargo commands that require it.

# Solution

Save the Noir ABI to disk when compiling from source, such that it can be used as a resource in a dApp. (Whether it is saved as a separate resource or incorporated into another is up for discussion.)

# Alternatives considered

None, access to the Noir ABI is a must have for any sort of e2e system.

### PR DESCRIPTION
**Pull request by TomAFrench**
# Related issue(s)

Resolves #897

# Description

## Summary of changes

Rather than serializing a `Circuit` into a `acir` file, we now serialize a `CompiledProgram` (`Circuit` + `Abi`) into a `json` file.

We still make use of the custom serialisation method for `Circuit` however JSON doesn't store bytes directly so we end up with an array. We could fix this by just encoding the entire thing into bytes rather than using json atm (and allowing exporting a JSON ABI separately).

An example output (for the `simple_shield` circuit) is shown below:

```

{
  "circuit": [0, 0, 0, 0, 21, 0, 0, 0, 3, 0, 0, 0, 2, 0, 0, 0, 13,  /*snip*/ ],
  "abi": {
    "parameters": [
      {
        "name": "priv_key",
        "type": {
          "kind": "field"
        },
        "visibility": "private"
      },
      {
        "name": "note_root",
        "type": {
          "kind": "field"
        },
        "visibility": "public"
      },
      {
        "name": "index",
        "type": {
          "kind": "field"
        },
        "visibility": "private"
      },
      {
        "name": "note_hash_path",
        "type": {
          "kind": "array",
          "length": 3,
          "type": {
            "kind": "field"
          }
        },
        "visibility": "private"
      },
      {
        "name": "to_pubkey_x",
        "type": {
          "kind": "field"
        },
        "visibility": "private"
      },
      {
        "name": "to_pubkey_y",
        "type": {
          "kind": "field"
        },
        "visibility": "private"
      }
    ],
    "param_witnesses": {
      "index": [
        3
      ],
      "note_hash_path": [
        4,
        5,
        6
      ],
      "note_root": [
        2
      ],
      "priv_key": [
        1
      ],
      "to_pubkey_x": [
        7
      ],
      "to_pubkey_y": [
        8
      ]
    },
    "return_type": {
      "kind": "array",
      "length": 2,
      "type": {
        "kind": "field"
      }
    },
    "return_witnesses": [
      13,
      15
    ]
  }
}
```

## Test additions / changes

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.
- [x] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

File: crates/nargo/src/cli/fs/keys.rs [hunk 1]
```diff
@@ -31,9 +31,9 @@ pub(crate) fn fetch_pk_and_vk<P: AsRef<Path>>(
     prove_circuit: bool,
     check_proof: bool,
 ) -> Result<(Vec<u8>, Vec<u8>), CliError> {
-    let mut acir_hash_path = PathBuf::new();
-    acir_hash_path.push(circuit_build_path.as_ref());
-    acir_hash_path.set_extension(ACIR_EXT.to_owned() + ".sha256");
+    let mut acir_hash_path = PathBuf::from(circuit_build_path.as_ref());
+    acir_hash_path.set_extension("json.sha256");
```
Comment (kevaundray) [inline crates/nargo/src/cli/fs/keys.rs:35]: Not a blocker as its unrelated; an extension of .checksum would be better here as we don't care for it being sha256

File: crates/nargo/src/cli/fs/program.rs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+) -> PathBuf {
+    let mut circuit_path = create_named_dir(circuit_dir.as_ref(), "target");
+    circuit_path.push(circuit_name);
+    circuit_path.set_extension("json");
+
+    write_to_file(&serde_json::to_vec(compiled_program).unwrap(), &circuit_path);
+
+    // Save a checksum of the circuit to compare against during proving and verification
+    let acir_hash = hash_constraint_system(&compiled_program.circuit);
+    circuit_path.set_extension("json.sha256");
```
Comment (kevaundray) [inline crates/nargo/src/cli/fs/program.rs:23]: This seems duplicated?

Comment (TomAFrench) [reply crates/nargo/src/cli/fs/program.rs]: Duplicated in what way? The only other time we interact with this is to read it along with the pk/vk

Comment (kevaundray) [reply crates/nargo/src/cli/fs/program.rs]: Oh hmm its just the extension part; in fetch_pk_and_vk, we also do:

```
acir_hash_path.set_extension("json.checksum");
```

Not a blocker as it was there before

File: crates/noirc_driver/src/program.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    pub circuit: Circuit,
+    pub abi: noirc_abi::Abi,
+}
+
+fn serialize_circuit<S>(circuit: &Circuit, s: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let mut circuit_bytes: Vec<u8> = Vec::new();
+    circuit.write(&mut circuit_bytes).unwrap();
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply crates/noirc_driver/src/program.rs:17]: It's usually ad-hoc, so the maintainers of acvm can say "after 2 version bumps, we delete it" -- In the next bump, we can remove all deprecated items in ACVM and then version bump and make a PR to Noir

**Review by kevaundray — APPROVED** [review]
LGTM

**Pull request merged**